src/game.ml

```diff
@@ -54,13 +99,191 @@ module type SoldierLogic = sig
 end
 
 module Pawn : SoldierLogic = struct
+  (** Returns true if the previous move is en passant-able (i.e. moved
+      two from the previous y-position), and that the current coordinate
+      position of the pawn can move diagonally due to en passant rule *)
+
+  let check_en_passant_left
+      (curr_x, curr_y)
+      ((last_x_old, last_y_old), (last_x_new, last_y_new))
+      color
+      board =
+    let last_piece = board.(last_x_new).(last_y_new) in
+    let net_y = last_y_new - last_y_old in
+
+    if color = White then
+      if
+        (* Looking diagonally up-left *)
+        last_piece = Some (Black, Pawn)
+        && net_y = -2
+        && last_x_new - curr_x = -1
+        && board.(curr_x - 1).(curr_y + 1) = None
+      then true
+      else false
+    else if
+      (* Looking diagonally down-left *)
+      last_piece = Some (White, Pawn)
+      && net_y = 2
+      && last_x_new - curr_x = -1
+      && board.(curr_x - 1).(curr_y - 1) = None
+    then true
+    else false
+
+  let check_en_passant_right
+      (curr_x, curr_y)
+      ((last_x_old, last_y_old), (last_x_new, last_y_new))
+      color
+      board =
+    let last_piece = board.(last_x_new).(last_y_new) in
+    let net_y = last_y_new - last_y_old in
+
+    if color = White then
+      if
+        (* Looking diagonally up-right *)
+        last_piece = Some (Black, Pawn)
+        && net_y = -2
+        && last_x_new - curr_x = 1
+        && board.(curr_x - 1).(curr_y + 1) = None
+      then true
+      else false
+    else if
+      (* Looking diagonally down-right *)
+      last_piece = Some (White, Pawn)
+      && net_y = 2
+      && last_x_new - curr_x = 1
+      && board.(curr_x - 1).(curr_y - 1) = None
+    then true
+    else false
+
+  let is_valid_square_pawn
+      (curr_x, curr_y)
+      board
+      color
+      last_move
+      (pot_x, pot_y) : bool =
+    let net_x = pot_x - curr_x in
+    let net_y = pot_y - curr_y in
+    let left_en_passant_able =
+      check_en_passant_left (curr_x, curr_y) last_move color board
+    in
+    let right_en_passant_able =
+      check_en_passant_right (curr_x, curr_y) last_move color board
+    in
+    let basic_valid_square =
+      is_valid_square board color (pot_x, pot_y)
+    in
+
+    let check_conditions =
+      if color = White then
+        if
+          (* Go forward two *)
+          basic_valid_square && curr_y = 1 && net_y = 2 && net_x = 2
+          && board.(curr_x).(curr_y + 1) = None
+          && board.(curr_x).(curr_y + 2) = None
+        then true (* Go forward one *)
+        else if
+          basic_valid_square && net_y = 1 && net_x = 0
+          && board.(curr_x).(curr_y + 1) = None
+        then true (* Diagonal up-left -- no en passant *)
+        else if
+          basic_valid_square
+          && (not left_en_passant_able)
```

Comment:
Why are you separating diagonal movement with and without en passant? If potential move is diagonal and either the square is not empty or it is left_en_passant_able, then we can move diagonally.